### PR TITLE
Makes sabotaged things respect friendly fire setting

### LINKED ISSUE #327: Sabotaging spy (and their team) can damage own sabotaged SG if friendly fire is off

### PR DESCRIPTION
**Pull request by DexterHaslem**
This makes sentry gun and dispenser damage given/received check friendly fire is on in respect to the saboteur's team. fixes #327 

I cannot spell saboteur correctly

### CODE REVIEW

**Comment by DexterHaslem** [comment]
this might cover #273 as well, I will check if that explosion code checks the same gamerules code

**Comment by DexterHaslem** [comment]
oh great, this doesnt work for malicious sabotage of SG because no saboteur info is set in that case to distinguish them. I need to change that code to have a proper flag and still keep track of sab team first.. this is WIP

**Comment by squeek502** [comment]
I'm not really seeing how #327 was happening in the first place. It looks like the previous code was essentially, "if an SG is sabotaged, anyone can damage it," and [`IsSabotaged()` is a prereq of `IsMaliciouslySabotaged()`](https://github.com/fortressforever/fortressforever/blob/a4ba59abe2cfdef801c893155a95f0debbcf4c4e/dlls/ff/ff_buildableobject.cpp#L602-L617), which means that `IsSabotaged()` must include maliciously sabotaged SGs, right? So I'm confused. Did you confirm #327 was real?

File: game_shared/ff/ff_gamerules.cpp [hunk 1]
```diff
@@ -2087,27 +2087,45 @@ bool CFFGameRules::FCanTakeDamage( CBaseEntity *pVictim, CBaseEntity *pAttacker
 	// if it's a buildable, then we use the buildable's owner to perform the team checks etc. -> Defrag
 	CBasePlayer *pBuildableOwner = NULL;
 
+	bool isFriendlyFireOn = friendlyfire.GetInt() != 0;
```
Comment (squeek502) [inline game_shared/ff/ff_gamerules.cpp]: `GetBool()` would make more sense here

File: game_shared/ff/ff_gamerules.cpp [hunk 2]
```diff
@@ -2087,27 +2087,45 @@ bool CFFGameRules::FCanTakeDamage( CBaseEntity *pVictim, CBaseEntity *pAttacker
...
 	}
 	
 	if ( pVictim && pVictim->Classify() == CLASS_SENTRYGUN )
 	{
 		CFFSentryGun *pSentry = dynamic_cast <CFFSentryGun *> (pVictim);
 
 		// Allow team to kill their own SG if it is sabotaged
-		if (pSentry && pSentry->IsSabotaged())
-			return true;
+		if ( pSentry && pSentry->IsSabotaged() )
```
Comment (squeek502) [inline game_shared/ff/ff_gamerules.cpp]: Haven't thought this through fully, but it seems like it should be possible to condense most/all of these if blocks into a single block and treat all buildables the same as appropriate, right? Check if `Classify() == CLASS_SENTRYGUN || Classify() == CLASS_DISPENSER`, then just cast to `CFFBuildableObject` or w/e, and all the code inside the if could be the same.

**Review by squeek502 — CHANGES_REQUESTED** [review]

File: game_shared/ff/ff_gamerules.cpp (hunk 2 above)
Comment (DexterHaslem) [reply game_shared/ff/ff_gamerules.cpp]: yes, strongly agree. I would have made it shorter but i didnt spend the time. Ill try to

File: game_shared/ff/ff_gamerules.cpp (hunk 1 above)
Comment (DexterHaslem) [reply game_shared/ff/ff_gamerules.cpp]: doh!

**Comment by DexterHaslem** [comment]
It definitely exists (just sab a SG then shoot it), thats what i was trying to address. I assumed the sabotage properties would always be set on the buildable base class, which is not true, I think. I will re-test

**Comment by squeek502** [comment]
Where does FCanTakeDamage return false in that instance (when trying to damage your maliciously sabotaged SG)? Or is it something else that's affecting things?

**Comment by DexterHaslem** [comment]
The previous code always returned true if it was sabotaged, is that what you are asking?

**Comment by squeek502** [comment]
Yeah, but if that's true, shouldn't it be impossible for there to be a bug where a (maliciously) sabotaged SG couldn't be damaged? Maybe I'm misunderstanding what #327 is saying?

**Comment by DexterHaslem** [comment]
Indeed, it's saying it could be damaged by the sabotuer which arguably doesn't make sense when friendly fire is on. this really comes down to a gameplay decision:

do sabotaged buildings 'belong' to the sabotaging team? I would not be surprised if the original devs intentionally made them 'building versus the world' when sabd

**Comment by squeek502** [comment]
Ohhh, ok I get it. The spy (and their team) can damage the gun they have maliciously sabotaged when FF is off. The title of #327 ~~is~~ (edit: was) super confusing.

**Comment by DexterHaslem** [comment]
True that! I'm traveling and will finish this up after next week

**Comment by DexterHaslem** [comment]
I debugged through this, didnt understand how sabotaging works - `m_hSaboteur` is always set so disregard that. I will refactor the duplicated sentry/dispenser stuff then this should be good to go

**Comment by DexterHaslem** [comment]
just tested it , works more gooder. also learned you can safely pass NULL to dynamic cast, so i'm just doing that to check it's a buildable

**Review by squeek502 — APPROVED** [review]

**Pull request merged**